open-cluster-management/backlog#14603 add volsync (#2771)
open-cluster-management/rhacm-docs: services/volsync.adoc

[#volsync]
= Replicating persistent volumes with VolSync (Technology Preview)

> VolSync is a Kubernetes operator that enables asynchronous replication of persistent volumes within a cluster, or across clusters with storage types that are not otherwise compatible for replication. It uses the Container Storage Interface (CSI) to overcome the compatibility limitation. After deploying the VolSync operator in your {ocp} environment, you can leverage it to create and maintain backup versions of your persistent data.

Reviewer: ```suggestion
VolSync is a Kubernetes operator that enables asynchronous replication of persistent volumes within a cluster, or across clusters with storage types that are not otherwise compatible for replication. It uses the Container Storage Interface (CSI) to overcome the compatibility limitation. After deploying the VolSync operator in your {ocp} environment, you can leverage it to create and maintain copies of your persistent data.
```

. Deploy the required VolSync resources to the clusters.

.. Download the https://github.com/open-cluster-management/policy-collection/blob/main/community/CM-Configuration-Management/policy-persistent-data-management.yaml[policy-persistent-data-management] policy template from the GitHub repository. In its current format, this policy starts an installation of VolSync on the hub cluster. After it is installed on the hub cluster, it automatically installs the required VolSync components on all of the managed clusters of the hub that were created by {ocp-short}.
+
You can specify a different label to identify which clusters will have VolSync installed by modifying the following section of the policy file:
+
[source,yaml]
----
spec:
  clusterSelector:
    matchLabels:
      vendor: OpenShift
----

.. Apply the policy while you are logged in to your hub cluster by entering the following command:
+
----
oc -n <namespace> apply -f ./policy-persistent-data-management.yaml
----
+

> Replace `namespace` with the namespace of your hub cluster nad managed clusters.

Reviewer: ```suggestion
Replace `namespace` with the namespace of your hub cluster and managed clusters.
```

Reviewer 2: I think namespace really just needs to be a namespace on the hub cluster itself - the policy will be created in this namespace on the hub.  We should probably remove mention of managed clusters on this line.

> . If your default `StorageClass` is not a CSI configuration, change your default CSI configuration to your VolSync CSI instance. 

Reviewer: ```suggestion
. Your default `StorageClass` should refer to a CSI-based storage driver, so it may be necessary to change the default. 
```

Reviewer 2: Wondering if we should be clear here where we should be checking for these storageclass/volumesnapshotclass requirements - This should be done on the clusters where VolSync is intended to be used. So for example if the customer has deployed (via the policy) to all their managed clusters with the intention of using Volsync there (as a replicationsource or replicationdestination) then they need to do these steps on those managed clusters.

They would potentially also need to do this on the hub cluster if they intend to use volsync on the hub cluster itself.

+
* Your steps might resemble the following procedure for an Amazon Web Services environment:

.. Change your default `StorageClass` by entering the following commands: 
+
----
kubectl annotate sc/gp2 storageclass.kubernetes.io/is-default-class="false" --overwrite
kubectl annotate sc/gp2-csi storageclass.kubernetes.io/is-default-class="true" --overwrite
----

> .. Install a `VolumeSnapshotClass` by running the following command:

Reviewer: ```suggestion
.. If a corresponding `VolumeSnapshotClass` does not already exist, one may be created by running the following command:
```

---
apiVersion: snapshot.storage.k8s.io/v1beta1
kind: VolumeSnapshotClass
metadata:
  name: gp2-csi
driver: ebs.csi.aws.com
deletionPolicy: Delete
SNAPCLASS
----

.. Set `standard-csi` as your default `VolumeSnapshotClass` by running the following command:
+
----
kubectl annotate volumesnapshotclass/standard-csi snapshot.storage.kubernetes.io/is-default-class="true"
----
+
Your CSI is now configured on a Google Cloud Platform environment.

. For Rsync-based replication, configure custom resources on the source and destination clusters. The custom resources use the `address` value to connect the source to the destination, and the `sshKeys` to ensure that the transferred data is secure. **Note:** You must copy the values for `address` and `sshKeys` from the destination to the source, so configure the destination before you configure the source. 

.. Configure your destination.

... Add the following minimal custom resource information to the YAML file of your destination cluster:
+
[source,yaml]
----
---
apiVersion: volsync/v1alpha1
kind: ReplicationDestination
metadata:
  name: <destination_name>
  namespace: <destination_namespace>
spec:
  rsync:
    copyMethod: Snapshot

>     capacity: 10Gi

Reviewer: Note: This capacity should match the capacity of the PVC that is being replicated

    accessModes: ["ReadWriteOnce"]
----
+
Replace `destination_name` with the name of your destination volume.
+
Replace `destination_namespace` with the name of the namespace where your destination is located.

... After the `ReplicationDestination` custom resource is created, copy the value of `.status.rsync.address`. This value is automatically generated when the custom resource is created. You will have to enter it when you configure the source. The following example shows the custom resource information that is added to the `ReplicationDestination` after it is created:
+
[source,yaml]
----
...
status:
  rsync:
    address: 10.01.101.001
    sshKeys: volsync-rsync-dest-src-test
----

> ... Copy the name of the secret that is provided as the value of `.status.rsync.sshKeys`. You will have to enter it when you configure the source.  

Reviewer: In addition to the name of the Secret, the Secret itself must be copied to the source cluster.

kubectl annotate sc/gp2 storageclass.kubernetes.io/is-default-class="false" --overwrite
kubectl annotate sc/gp2-csi storageclass.kubernetes.io/is-default-class="true" --overwrite
----

.. If a corresponding `VolumeSnapshotClass` does not exist, create one by running the following command:
+
----
kubectl create -f - << SNAPCLASS
---
apiVersion: snapshot.storage.k8s.io/v1beta1
kind: VolumeSnapshotClass
metadata:
  name: gp2-csi
driver: ebs.csi.aws.com
deletionPolicy: Delete
SNAPCLASS
----

.. Set `gp2-csi` as your default `VolumeSnapshotClass` by running the following command:
+
----
kubectl annotate volumesnapshotclass/gp2-csi snapshot.storage.kubernetes.io/is-default-class="true"
----
+
Your CSI is now configured on an Amazon Web Services environment.

* The steps might resemble the following for an Google Cloud Platform environment:

.. Change your default `StorageClass` by entering the following commands: 
+
----
kubectl annotate sc/standard storageclass.kubernetes.io/is-default-class="false" --overwrite
kubectl annotate sc/standard-csi storageclass.kubernetes.io/is-default-class="true" --overwrite
----

.. Install a `VolumeSnapshotClass` by running the following command:
+
----

> kubectl create -f - << SNAPCLASS
> ---
> apiVersion: snapshot.storage.k8s.io/v1beta1
> kind: VolumeSnapshotClass
> metadata:
>   name: gp2-csi
> driver: ebs.csi.aws.com
> deletionPolicy: Delete
> SNAPCLASS

Reviewer: Looks like this snapclass is for AWS, not GCP. Here's one for GCP:

```suggestion
kubectl create -f - << SNAPCLASS
---
apiVersion: snapshot.storage.k8s.io/v1beta1
kind: VolumeSnapshotClass
metadata:
  name: standard-csi
driver: pd.csi.storage.gke.io
deletionPolicy: Delete
SNAPCLASS
```

[#volsync]
= Replicating persistent volumes with VolSync (Technology Preview)

> VolSync is a Kubernetes operator that enables asynchronous replication of persistent volumes within a cluster, or across clusters with storage types that are not otherwise compatible for replication. It uses the Container Storage Interface (CSI) to overcome the compatibility limitation. After deploying the VolSync operator in your {ocp} environment where you have your {product-title} hub cluster, you can leverage it to create and maintain backup versions of your {product-title-short} managed clusters.

Reviewer: - I don't think we're advocating always installing on the hub, just on clusters that are source or destination.
- Original phrasing makes it sound like we're backing up the cluster instead of just the data

```suggestion
VolSync is a Kubernetes operator that enables asynchronous replication of persistent volumes within a cluster, or across clusters with storage types that are not otherwise compatible for replication. It uses the Container Storage Interface (CSI) to overcome the compatibility limitation. After deploying the VolSync operator in your {ocp} environment, you can leverage it to create and maintain backup versions of your persistent data.
```

**Note:** VolSync does not meet the requirements of the FIPS standard. 

There are three methods that you can use to replicate when you use VolSync, which depend on the number of synchronization locations that you have. The Rsync method is used for this example. For information about the other methods and more information about Rsync, see https://volsync.readthedocs.io/en/latest/usage/index.html[Usage] in the VolSync documentation.  

> * Rsync replication is a one-to-one replication of persistent volumes, and is likely to be the most commonly used. This is used for typical backups at a remote site or mirroring an environment. 

Reviewer: ```suggestion
* Rsync replication is a one-to-one replication of persistent volumes, and is likely to be the most commonly used. This is used for replicating data to a remote site. 
```
(removing "backup" wording to avoid confusion w/ the restic-based mover which is specifically for backup)

[#volsync-prereq]
== Prerequisites

Before installing VolSync on your clusters, you must have the following requirements:

* A configured {ocp} environment running a {product-title-short} version 2.4, or later, hub cluster.

* At least two configured clusters that are managed by the same {product-title-short} hub cluster.

[#volsync-install-clusters]
== Installing VolSync on the managed clusters

To enable VolSync on two clusters in your environment, you must install it on both the source and the target managed clusters. Complete the following steps to leverage the policy function in {product-title-short} to install VolSync:

. Deploy the required VolSync resources to the clusters.

> 

Reviewer 2: I've discussed a bit with @jnpacker and I think to install volsync, rather than being a guided experience via deploying an application, the user would instead need to create the policy using the sample one here:

https://github.com/open-cluster-management/policy-collection/blob/main/community/CM-Configuration-Management/policy-persistent-data-management.yaml

If they create this policy on the hub cluster it should then create a volsync subscription on the hub which then targets managed clusters and deploys volsync to them

Author: Discussed with Tesshu, and updated this procedure.

.. In the {product-title-short} console navigation, select *Applications*. 

.. Select *Create application* to start a new application. The application that you create uses an existing policy to install the VolSync resources.

.. Complete the required information for your application.

> 

Reviewer 2: When I try this, I have to pick `Subscription` as the type in the dropdown before I can continue

Author: Changed the procedure.

... Enter `policy-persistent-data-management` as the name of your application. This name corresponds to the policy that is created in the Git repository that deploys the VolSync content on all of your clusters that are managed by the hub. 

... Select the option for a Git repository, so you can reference the policy file.

... Use the following URL for the Git repository: \https://github.com/open-cluster-management/policy-collection.git.

... Select the `main` branch. 

... Add `stable/CM-Configuration-Management` for the path.

... Make sure you select the option to *Deploy to all online clusters and local-cluster*. This ensures that it is included on all of your managed clusters. 

.. Select `Save` to save your application. Within a few minutes, the application will start deploying to all of the clusters that are managed by your {product-title-short} hub cluster. 

. Change your default CSI configuration to your VolSync CSI instance. 
+
* Your steps might resemble the following procedure for an Amazon Web Services environment:

> ... Stop the current service: 

Reviewer: The stop/start terminology doesn't quite match what's actually going on...
We're changing which StorageClass is marked as the default one (i.e., if the user doesn't specify a specific StorageClass in their PVC, it will be assigned to the SC that is currently marked as default.)

Same goes for the VolumeSnapshotClass explanation below

+
----
kubectl annotate sc/gp2 storageclass.kubernetes.io/is-default-class="false" --overwrite
----

... Redirect the service to CSI and start the service: 
+
----
kubectl annotate sc/gp2-csi storageclass.kubernetes.io/is-default-class="true" --overwrite
----

> ... Install a `VolumeSnapshotClass` by running the following command:

Reviewer: I believe OCP 4.9 has started shipping w/ a default VolumeSnapshotClass that is compatible w/ the ebs csi driver, so these instructions may not be necessary depending on the OCP version.

----

... Install a `VolumeSnapshotClass` by running the following command:
+
----
kubectl create -f - << SNAPCLASS
---
apiVersion: snapshot.storage.k8s.io/v1beta1
kind: VolumeSnapshotClass
metadata:
  name: gp2-csi
driver: ebs.csi.aws.com
deletionPolicy: Delete
SNAPCLASS
----

... Set `standard-csi` as your default `VolumeSnapshotClass` by running the following command:
+
----
kubectl annotate volumesnapshotclass/standard-csi snapshot.storage.kubernetes.io/is-default-class="true"
----
+
Your CSI is now configured on a Google Cloud Platform environment.

. For Rsync-based replication, configure a custom resource on the source and target clusters.

.. Add the following minimal custom resource information to the YAML file of your source cluster:
+
[source,yaml]
----
---
apiVersion: volsync.backube/v1alpha1
kind: ReplicationSource
metadata:
  name: <source_name>
  namespace: <source_namespace>
spec:
  sourcePVC: <persistent_volume_claim>
  trigger:
    schedule: "*/5 * * * *"
  rsync:
    sshKeys: <volsync-rsync-destination-src-database-destination>
    address: <source.host.com>
    copyMethod: Clone
----
+

> Replace `source_name` with the name of your source volume.

Reviewer: This can be an arbitrary name, unrelated to the volume (PVC) name

+
Replace `source_namespace` with the name of the namespace where your source is located.
+
Replace `persistent_volume_claim` with the name of your source claim.
+

> Replace `source.host.com` with the host address of your source cluster. 

Reviewer: The `sshKeys` and `address` fields need to be taken from the ReplicationDestination.

I suggest re-arranging so that the user creates the destination before the source.
Upstream rsync mover usage: https://volsync.readthedocs.io/en/latest/usage/rsync/index.html

Once the ReplicationDestination is created, the operator will populate its `.status.rsync.address` field, and that info needs to be copied here.

The ReplicationDestination will also write `.status.rsync.sshKeys` which is the name of a Secret that the operator has created. This Secret needs to be exported from the destination cluster and imported into the source cluster in the same namespace as the ReplicationSource object. The name of this imported Secret should then be provided in the ReplicationSource's `.spec.rsync.sshKeys` field.

Author: Thank you for the clarification. I rearranged this procedure.

.. Add the following minimal custom resource information to the YAML file of your target cluster:
+
[source,yaml]
----
---
apiVersion: volsync/v1alpha1
kind: ReplicationDestination
metadata:
  name: <destination_name>
  namespace: <destination_namespace>
spec:
  rsync:
    copyMethod: Snapshot
    capacity: 10Gi
    accessModes: ["ReadWriteOnce"]
----
+
Replace `destination_name` with the name of your destination volume.
+
Replace `destination_namespace` with the name of the namespace where your destination is located.

Your can now set up the synchronization method of the persistent volume.

[#volsync-start]
== Starting your synchronization

You have a few options to select from when determining how you start your replications: always running, on a schedule, or manually. Implement the method from the following list that best meets your needs:  

> * *Always:* Runs a constant replication. When a change is detected on the source persistent volume, a replication is automatically started. Your setting for this replication option might resemble the following content:

Reviewer: I suggest we leave out the "always" docs. We really want to guide users toward scheduled replication for most use cases.
